enclave/core/sgx/enter.S

```diff
@@ -160,21 +170,25 @@ oe_enter:
 #define OM_HOST_RBP                 (-2*8)(%rbp)
 #define OM_HOST_OUTPUT_ARG1         (-3*8)(%rbp)
 #define OM_HOST_OUTPUT_ARG2         (-4*8)(%rbp)
-#define OM_HOST_RETURN_ADDR         (-5*8)(%rbp)
+#define OM_ENC_TD                   (-5*8)(%rbp)
+#define OM_HOST_RETURN_ADDR         (-6*8)(%rbp)
 
     // Allocate stack.
     sub $OM_STACK_LENGTH, %rsp
 
     // Save the host stack pointers to enclave stack.
-    mov %fs:td_host_rsp, %r8
-    mov %fs:td_host_rbp, %r9
+    mov td_host_rsp(%r11), %r8
+    mov td_host_rbp(%r11), %r9
     mov %r8, OM_HOST_RSP
     mov %r9, OM_HOST_RBP
 
     // Save the host return address to enclave stack.
-    mov %fs:td_host_rcx, %r8
+    mov td_host_rcx(%r11), %r8
     mov %r8, OM_HOST_RETURN_ADDR
 
+    // Save reference to the td structure to enclave stack.
+    mov %r11, OM_ENC_TD
+
     // Call __oe_handle_main(ARG1=RDI, ARG2=RSI, CSSA=RDX, TCS=RCX, OUTPUTARG1=R8, OUTPUTARG2=R9)
```

Comment:
@anakrish doesn't `__oe_handle_main` never return? (Instead it calls oe_asm_exit eventually). I am wondering why this function has any code after calling `__oe_handle_main`. Am I misunderstanding something?

If this is the case, we could just clear all the registers and call `ud2` just as a precaution.

Reply:
`__oe_handle_main` will return when the ecall completes. `oe_asm_exit` is used for ocalls.
There is some code that is the same between the two and I might have a branch that tries to avoid the duplication.